### ISSUE TITLE
EXP-003 research: synthetic counterfactual typed-relation learning

### ISSUE THREAD

**Issue by gr-rr** [ListenCloser/listencloser#1306]
Program owner: #1303.
Research synthesis: #1297.
Related product owners: #812 (Similar Moments), #1175 (Structure Map), #1233 (Melody), #1300 (bespoke cleanup).
Optional representation input: #1304.

## Question

> **Can known symbolic music transformations manufacture enough reliable supervision to learn useful typed musical relations essentially for free?**

This is the most novel first-wave pilot, but it must remain cheap and mechanically verifiable.

## Core hypothesis

Music gives us transformations whose semantic effects are partially known by construction.

For an exact symbolic passage `x`, create transformed passage `T(x)` and label what should be preserved or changed.

Examples:

| Transformation | Expected preserved | Expected changed |
|---|---|---|
| transpose | interval contour, rhythm, broad structure | absolute pitch/key |
| tempo scale | relative rhythm/order, pitch relations | absolute timing |
| octave shift | pitch-class/interval pattern, rhythm | register |
| instrument remap | symbolic content | instrumentation metadata/rendered timbre |
| controlled note deletion | most global context if sparse | local content/density |
| ornament insertion | coarse skeleton/phrase identity | surface note sequence |
| simple revoicing/inversion | some harmonic identity/rhythm | register/voicing/texture |

The research question is whether frozen representations and a tiny relation model can recover these **typed relations**, rather than collapsing everything into one generic similarity score.

## First-pass scope

Start symbolic-only. No audio rendering is required for the first result.

Use rights-safe/public MIDI/score material already available through existing evaluation datasets or a small deterministic public-domain corpus. Do not create a new dataset acquisition program.

Generate a fixed synthetic pair manifest:

```text
source_id
source_hash
transform_type
transform_parameters
transformed_hash
known_preserved_attributes
known_changed_attributes
split
```

Freeze the transformation policy and split before the result-bearing run.

## Tooling

Prefer mature OSS:
- pretty_midi
- music21
- Partitura
- NumPy/PyTorch/scikit-learn
- frozen M2BERT/MidiBERT embeddings only if lawful/reproducible and useful

Do not write a new symbolic transformation framework. Keep transformations as small deterministic functions with round-trip/invariant tests.

## Phase 1 — representation-only diagnostics

Before training a relation classifier, measure how a frozen representation behaves under each transformation.

For passage embedding function `f`:

```text
cosine(f(x), f(T(x)))
nearest-neighbor retrieval of source under transform
layer-wise behavior if cheap
```

This establishes whether existing representations already encode the desired invariances/equivariances.

## Phase 2 — tiny typed-relation head

Train a small classifier/ranker over pairs:

```text
[f(x), f(T(x)), |f(x)-f(T(x))|]
→ relation labels
```

Initial labels should stay concrete, e.g.:
- same_interval_melody_pattern
- same_rhythm_pattern
- transposed
- tempo_scaled
- register_changed
- surface_ornamented
- revoiced

Do not claim human-level motif/form semantics from synthetic transforms.

## Negative / hard controls

Include mismatched random passage pairs and, where possible, transformations that deliberately break one relation while preserving another.

Example:

```text
same melody / altered rhythm
vs
same rhythm / altered melody
```

This is necessary to test disentanglement rather than trivial pair identity.

## Metrics

Primary:
- typed-relation classification macro-F1 or accuracy on a held-out source-piece split.

Secondary:
- Recall@K for retrieving the original/related passage under each transform;
- confusion matrix between relation types;
- invariance/equivariance diagnostics by transformation.

Split by **source piece**, never by generated pair, so transformed variants of one source cannot leak across train/test.

## Compute budget

Most work should be CPU-local.

```text
GPU-hours <= 4
cash cost <= $5
synthetic pair count: start small (thousands, not millions)
```

Use local CPU/MPS for transformation generation and small models; free Kaggle/Modal/RunPod only if frozen embedding extraction benefits materially.

Do not scale pair count merely because synthetic data is cheap. First prove the task is nontrivial and useful.

## Success criterion

A useful result must show more than “the model identifies transpose.”

Preferred success:

> On held-out source pieces, a tiny model can distinguish at least several typed preserve/change relationships and outperform simple handcrafted distance controls, with especially clear separation between melody-preserving/rhythm-changing and rhythm-preserving/melody-changing pairs.

Compare against at least one simple baseline, e.g. interval/rhythm descriptor distances or raw embedding cosine.

## Product interpretation if successful

A positive result may justify later work on:
- typed Similar Moments relations;
- cross-Work passage retrieval;
- motif/variation candidates;
- representation robustness testing;
- synthetic augmentation for #1304/#1305.

It does **not** directly justify changing product Similar Moments/Structure truth.

## Stop / kill criteria

Stop if:
- transformations cannot be made semantically clean enough to define honest labels;
- performance is explained almost entirely by trivial transform artifacts/metadata;
- simple handcrafted controls already solve the synthetic task and learned representations add no useful generalization;
- held-out source-piece performance collapses;
- useful results require large backbone fine-tuning outside the budget.

## Non-goals

- no audio foundation-model training;
- no million-example synthetic corpus;
- no generative-model training;
- no claim that synthetic relation labels equal human similarity judgments;
- no universal similarity score;
- no graph database;
- no production UI in this issue.

## Deliverables

1. Fixed public/right-safe source corpus manifest.
2. Deterministic tested transformation library limited to this experiment.
3. Generated pair manifest with source-piece-disjoint splits.
4. Frozen-representation diagnostics.
5. Simple baseline vs tiny relation-head results.
6. Per-transform confusion/error analysis.
7. Compute/runtime/rights record.
8. Explicit disposition:

```text
FAIL — synthetic supervision does not yield useful typed relations
NARROW — one/two transformations work, insufficient broader thesis
CONTINUE — several typed relations generalize and justify follow-on use
```

## Acceptance

- [ ] source-piece-disjoint split frozen;
- [ ] transformation semantics/invariants covered by tests;
- [ ] simple handcrafted/embedding baseline included;
- [ ] multiple typed relations evaluated, not only transformation ID;
- [ ] <=4 GPU-hours / <=$5 or stop with reason;
- [ ] result reproducible from exact manifest/seed/environment;
- [ ] FAIL/NARROW/CONTINUE written back to #1303;
- [ ] no product semantic promotion from synthetic-only evidence.

Refs #1303 #1304 #1305 #1297 #812 #1175 #1233 #1300.

**Comment by gr-rr** [comment]
## Kickoff / phase-0 refinement

This experiment is compatible with the no-manual-label constraint because the first labels are deterministic consequences of the transforms. Keep it that way.

### Phase 0 — no training

Before fitting a relation head, generate a small fixed diagnostic corpus:

```text
~200–500 source passages
source-piece-disjoint split
5 bounded transforms initially:
  transpose
  octave shift
  tempo scale
  melody/pitch-pattern perturbation
  rhythm perturbation
```

Sparse note deletion/ornamentation may be added only if the first transform semantics are clean.

For each frozen representation, first report:
- source↔transform cosine/distance distributions;
- original retrieval Recall@K under each transform;
- separation between melody-preserved/rhythm-changed and rhythm-preserved/melody-changed controls.

If these diagnostics are trivial, collapsed, or explained by transform artifacts, stop before training.

### Phase 1 only if earned

Then create thousands, not millions, of deterministic pairs and train one tiny typed-relation classifier against handcrafted descriptor/distance controls.

### M2BERT compute hygiene

Reuse #1304's frozen extractor/cache. Avoid padding every short synthetic passage to 1024 notes: batch/bucket at bounded lengths such as 128/256/512/1024 with masks where the upstream model contract supports it. Cache only the predeclared layer(s), preferably FP16.

Local CPU owns transformation generation/invariant tests and tiny-head training. #1311 Kaggle T4 is only for bounded frozen embedding materialization if useful.

### LLM policy

An LLM may help generate *candidate transformation ideas, names, or adversarial test cases*, but it is not the semantic oracle. Every result-bearing relation label must be deterministic by construction or backed by an external reference.

Refs #1303 #1304 #1305 #1311 #812 #1175.

**Comment by gr-rr** [comment]
Execution kickoff: branch `research/1306-synthetic-relations` created from current main (`d11f0c46...`). Phase 1 may proceed in parallel with #1304/#1305: deterministic transformations + source-piece-disjoint manifest + handcrafted descriptor / frozen-representation diagnostics only. Do **not** start with a trained relation model or huge synthetic corpus. Prefer music21 / pretty_midi / Partitura transformation primitives and tests; no custom transformation framework. If M2BERT embeddings become useful, consume #1304's frozen cache rather than recomputing them independently. Kaggle CPU is sufficient for generation/diagnostics; use T4 only for expensive frozen embedding materialization.

**Comment by gr-rr** [comment]
Execution update — Phase 0 handcrafted control completed on free GitHub CPU before spending any representation/GPU compute.

Setup:
- fixed POP909 revision `d83e6edba6872a704f5d3b8b32f5cb540088dae6`
- 30 source pieces, source-piece-disjoint split: 18 train / 6 valid / 6 test
- 6 deterministic transformations: transpose +5, octave shift, tempo-scale 1.25, melody perturbation with timing fixed, rhythm perturbation with pitch fixed, delete every 5th note
- labels are 5 preserve/change relations, not transform IDs
- baseline uses only 7 handcrafted pair descriptors (pitch-class/interval/rhythm/duration histogram distances, mean-pitch delta, span ratio, note-count ratio) + sklearn logistic regression
- no GPU / no learned music representation

Held-out result:
- validation macro-F1: **0.8209**
- test macro-F1: **0.8306**
- test per relation: interval-pattern-preserved 0.857; rhythm-pattern-preserved 0.829; register-changed 1.000; pitch-content-changed 0.800; density-changed 0.667.

Interpretation: this is useful mostly as an adversarial control. The current synthetic task is already substantially solvable by trivial handcrafted statistics, so `identify deterministic transform/preserve-change relation` is **not sufficient evidence** for a learned representation. Do not jump straight to Phase 2.

Disposition for this slice: **NARROW / refine task**. Next step should add hard counterfactuals deliberately matched on these simple descriptors (especially same melody / altered rhythm vs same rhythm / altered melody, plus random-source hard negatives) and only then compare frozen M2BERT embeddings against the handcrafted control. If the simple baseline remains near this strong after hardening, stop rather than training a relation model.
